### PR TITLE
Fix infinite introspection loop

### PR DESCRIPTION
**Pull request by hasparus**
`fetcher` was depending on `schema`, but we did not set the `schema` prop on `GraphiQLProvider`, so for each new fetcher we received a new (same but not referentially equal) schema, which created a new fetcher (same but not referentially equal), what resulted in an infinite loop of rerenders and introspection queries.

---

## Before

https://user-images.githubusercontent.com/15332326/231020797-04c84f67-baf3-45ad-8674-eba336371e7e.mov

## After

https://user-images.githubusercontent.com/15332326/231020794-64177768-3622-48e6-98fc-f7315f7a7c73.mov

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -112,7 +115,7 @@ export function GraphProtocolGraphiQL<TQuery extends SavedQuery>({
   const explorerPlugin = useExplorerPlugin({
     query: querySource,
     onEdit: setQuerySource,
-    storage,
+    showAttribution: false,
```
Comment (hasparus) [inline src/index.tsx:118]: This is unrelated, but TypeScript told me `storage` is no longer accepted, and `showAttribution` is required.

File: src/index.tsx [hunk 2]
```diff
@@ -121,6 +124,7 @@ export function GraphProtocolGraphiQL<TQuery extends SavedQuery>({
       query={querySource}
       storage={storage}
       plugins={[explorerPlugin]}
+      schema={schema}
```
Comment (hasparus) [inline src/index.tsx:127]: We pass `schema` here, firstly with value `undefined` to prompt introspection, then with the schema we received to ensure that `fetcher` change will not cause another introspection.

(`null` would say "we don't want introspection")

Ideally, we'd clear the schema after `fetcherOptions.url` change, but I'm lazy and we won't need that in foreseeable future.

File: demo/index.tsx [hunk 3]
```diff
@@ -46,6 +46,8 @@ const initialQueries = [
   ...x,
 }))
 
+const fetcherOptions = { url }
```
Comment (hasparus) [inline demo/index.tsx:49]: This had to be moved so it doesn't fire recomputing of the `useMemo` with `extendFetcherWithValidations`. I did the same thing in Studio.

File: src/index.tsx [hunk 4]
```diff
@@ -92,14 +92,17 @@ export function GraphProtocolGraphiQL<TQuery extends SavedQuery>({
   className,
   graphqlValidations = true,
 }: GraphProtocolGraphiQLProps<TQuery>) {
-  const [schema, setSchema] = useState<GraphQLSchema | null>(null)
+  const [schema, setSchema] = useState<GraphQLSchema | undefined>(
+    // `undefined` will trigger introspection for the schema when given to `GraphiQLProvider`
```
Comment (saihaj) [inline src/index.tsx:96]: so `null` doesn't trigger Introspection?

**Review by saihaj — APPROVED** [review]

**Review by dotansimha — APPROVED** [review]

File: src/index.tsx (hunk 4 above)
Comment (hasparus) [reply src/index.tsx:96]: Yup

https://github.com/graphql/graphiql/blob/a8f21ad3cf1c2ead95fa2c95372d01bafff8fee9/packages/graphiql-react/src/schema.tsx#L97-L113

**Comment by hasparus** [comment]
I've found another small bug.
We've accidentally bundled `graphql` what results in "Cannot use GraphQLSchema from another realm".

image

I've added `graphql` to peer deps to fix it.

Could you guys (@dotansimha @saihaj) maybe recommend me a build tool that doesn't bundle, but still emits both cjs and esm? I feel like bundling is the most common / status quo solution in the community, but it's very rarely something that people want or expect. Is [Bob](https://github.com/kamilkisiela/bob) meant for public use?

### Now

image

### Before

image

**Pull request merged**